src/storage/cache_proxy/cluster_manager.rs

```diff
@@ -0,0 +1,1096 @@
+//! The utilities of distribute cache cluster management
+
+use std::sync::Arc;
+use std::time::Duration;
+
+use anyhow::Context;
+use arc_swap::{ArcSwap, ArcSwapOption};
+use futures::StreamExt;
+use tracing::{debug, warn};
+
+use crate::async_fuse::memfs::kv_engine::etcd_impl::Session;
+use crate::async_fuse::memfs::kv_engine::{KVEngine, SetOption};
+use crate::async_fuse::memfs::kv_engine::{KVEngineType, KeyType, ValueType};
+use crate::common::error::{DatenLordError, DatenLordResult};
+
+use super::node::{MasterNodeInfo, Node, NodeStatus};
+use super::ring::Ring;
+
+/// The timeout for current node's session
+const SESSION_TIMEOUT_SEC: u64 = 10;
+
+/// This struct is used to interact with etcd server and manager the cluster.
+#[derive(Debug, Clone)]
+#[allow(dead_code)]
+pub struct ClusterManager {
+    /// inner struct
+    inner: Arc<ClusterManagerInner>,
+}
+
+#[allow(dead_code)]
+impl ClusterManager {
+    /// Create new cluster manager
+    pub fn new(kv_engine: Arc<KVEngineType>, node: Node) -> Self {
+        let inner = Arc::new(ClusterManagerInner::new(kv_engine, node));
+        Self { inner }
+    }
+
+    /// Get current node
+    #[must_use]
+    pub fn get_node(&self) -> Node {
+        self.inner.get_node()
+    }
+
+    /// Get current hashring
+    pub async fn get_ring(&self) -> DatenLordResult<Ring<Node>> {
+        self.inner.get_ring(false).await
+    }
+
+    /// Get all nodes
+    pub async fn get_nodes(&self) -> DatenLordResult<Vec<Node>> {
+        self.inner.get_nodes(false).await
+    }
+
+    ///  Stop the cluster manager
+    pub fn stop(&self) {
+        // In this step, we just to clean session and force down the cluster manager
+        self.inner.clean_sessions();
+    }
+
+    /// Run the cluster manager as state machine
+    ///
+    /// We need to perpare current node info, current node list and generate hash ring info
+    pub async fn run(&self) -> DatenLordResult<()> {
+        loop {
+            // 0. Prepare for the session
+            self.inner.prepare().await?;
+
+            // 1. Init cluster manager
+            debug!(
+                "Cluster manager start to run in node: {:?}",
+                self.inner.get_node().ip()
+            );
+            let mut current_node_info = self.inner.get_node();
+            // Next step is to register the node
+            current_node_info.set_status(NodeStatus::Registering);
+            self.inner.update_node(current_node_info);
+            self.inner.update_node_in_cluster().await?;
```

Comment:
set_status(), update_node(), update_node_in_cluster() always call together? 
should be merge into one

Reply:
`set_status(NodeStatus::Registering);` may be unnecessary

Reply:
Master does not need to know each node's status.  Thus removing `status` from Node struct, the code will be much simpler

Reply:
Yes, the values implemented here are to make it easier to understand the process, and I've removed duplicate steps.

Reply:
However, NodeStatus is still required for local processing, and here I still made a reservation to facilitate the handling of role-related logic, and skipped NodeStatus during sequence talk to reduce the consistency maintenance of role status.

Reply:
At this stage, NodeStatus may indeed not be necessary.